### PR TITLE
[WIP] Add Flax LoRA Support to Dreambooth

### PR DESCRIPTION
**Pull request by yasyf**
I saw @patrickvonplaten is working on LoRA support for the non-Flax Dreambooth in #1884. We've been taking a stab at implementing LoRA support for TPUs, taking example from the patching method used by @cloneofsimo in [cloneofsimo/lora](https://github.com/cloneofsimo/lora).

I've got it successfully patching and training, but the output is currently no good. I'm reaching the end of the time I have allocated for this—might pick it up in the future, but for now I'm putting this up in case anyone finds it useful!

### CODE REVIEW

**Comment by Thomas-MMJ** [comment]
are you using the higher learning rate (1e-4) and not the typical dreambooth rate (1e-6)?

**Comment by yasyf** [comment]
@Thomas-MMJ good call, let me try that

**Comment by yasyf** [comment]
yea I still get `NaN`s for the loss function right away. Will leave in case its useful for @patrickvonplaten

**Comment by yasyf** [comment]
@cloneofsimo any ideas?

**Comment by Thomas-MMJ** [comment]
@yasyf is your base model 16 or 32 bit? See the recent discussion that half can result in NaNs.

**Comment by yasyf** [comment]
Made some progress. I was only patching `FlaxAttentionBlock`. By adding `FlaxGEGLU` to the list, I'm now able to get it to start training. However, after a few steps, the loss rapidly blows up, so this still isn't usable :(

```
Epoch... (1/100 | Loss: 0.22375227510929108)
Epoch... (2/100 | Loss: 0.42204561829566956)
Epoch... (3/100 | Loss: 0.9007828235626221)
Epoch... (4/100 | Loss: 1.4756149053573608)
Epoch... (5/100 | Loss: 344.7566223144531)
Epoch... (6/100 | Loss: nan)
```

@Thomas-MMJ nope!

**Comment by krahnikblis** [comment]
@yasyf are you training the _lora up/down weights_ directly, or taking low-rank approximation of the _model gradients_? 

i've also reinvented this wheel (building LoRA on Flax for Diffusers), but i'm a purist; i built the model as a whole separate entity, which acts on the params of the unet and text encoder, prior to inference - meaning the diffusers models themselves are not altered (no replacing/wrapping model layers), only their parameters. this works really well for extracting from all the fine-tune models, and combining all those extractions together into a single lora... 

so in training, my thoughts were to use the dreambooth script, get the gradients against the text encoder and unet models' params, then take the _low-rank approximation of the grads_, and those are the same shape as the lora layers, so the apply_gradients is done on a trainstate of the lora model. it's a bit more mathsy but essentially i thought that compressing the important parts of the train step gradients and then applying those to the lora params, it would actually train a lot faster due to higher learn rates enabled by the SVD filtering of noise... but, i can't get the dang thing to compile (and on expensive TPU colab, i don't let it try longer than a few minutes), and if it does (e.g. after 25 mins on CPU colab), it OOMs when training starts. might just be i don't know jax, or might be that i'm trying to apply gradients to a different set of params than optax calculates them on, complicating all the tracing it's doing under its magic little hood... 

it would definitely simplify things to use the lora up/down params themselves as the gradient descent target (calculate loss and grads using the lora up/down params as the train state), but i thought that would loose the ability to leverage lora's strength (or rather, SVD's strength) in filtering the "unimportant" parts of a matrix (the matrix being the gradients).

**Comment by yasyf** [comment]
@krahnikblis if you send me the code I can run it on a beefy v4 TPU to see if memory is the issue
